app/src/main/java/com/fivegmag/a5gmsmediasessionhandler/MediaSessionHandlerMessengerService.kt

```diff
@@ -302,5 +323,177 @@ class MediaSessionHandlerMessengerService() : Service() {
+                "Entry[${dataReporting?.mediaPlayerEntry}], " +
+                "startTime[${dataReporting?.consumptionReportingUnits?.get(0)?.startTime}]," +
+                "duration[${dataReporting?.consumptionReportingUnits?.get(0)?.duration}]," +
+                "ipAddr[${dataReporting?.consumptionReportingUnits?.get(0)?.mediaEndpointAddress?.ipv4Addr}/" +
+                "${dataReporting?.consumptionReportingUnits?.get(0)?.mediaEndpointAddress?.ipv6Addr}]"
+        )
+
+        Toast.makeText(
+            applicationContext,
+            "MSH recv Consumption-ID: ${dataReporting?.reportingClientId}",
+            Toast.LENGTH_LONG
+        ).show()
+
+        // call m5 report consumption to AF - TS26.512 Clause 4.7.4
+        val consumptionReportingApi: ConsumptionReportingApi? = clientsSessionData[sendingUid]?.consumptionReportingApi
+        if(consumptionReportingApi == null)
+        {
+            Log.i(TAG, "[ConsumptionReporting] consumptionReportingApi is 【NULL】")
+            return
+        }
+
+        val provisisioningSessionId: String = "2";
+        val call: Call<ResponseBody>? = consumptionReportingApi.postConsumptionReporting(provisisioningSessionId, dataReporting?.reportingClientId);
+
+        call?.enqueue(object : retrofit2.Callback<ResponseBody> {
+            override fun onResponse(
+                call: Call<ResponseBody?>,
+                response: Response<ResponseBody?>
+            ) {
+                Log.i(TAG, "[ConsumptionReporting] resp from AF:" + response.body()?.string())
+            }
+
+            override fun onFailure(call: Call<ResponseBody?>, t: Throwable) {
+                Log.i(TAG, "[ConsumptionReporting] onFailure")
+                call.cancel()
+            }
+        })
+    }
+    
+    private fun createRetrofitForConsumpReport(clientId: Int) {
+        try {
+            var consumpReportUrl: String = "";
+            if(clientsSessionData[clientId]?.serviceAccessInformation!!.clientConsumptionReportingConfiguration.serverAddresses.isNotEmpty())
+            {
+                consumpReportUrl = clientsSessionData[clientId]?.serviceAccessInformation!!.clientConsumptionReportingConfiguration.serverAddresses[0];
+            }
+            Log.i(TAG, "[ConsumptionReporting] createRetrofitForConsumpReport URL: $consumpReportUrl.")
+
+            if (consumpReportUrl != "") {
+                val retrofit = retrofitBuilder
+                    .baseUrl(consumpReportUrl)
+                    .build()
+
+                clientsSessionData[clientId]?.consumptionReportingApi = retrofit.create(ConsumptionReportingApi::class.java)
+            }
+        } catch (e: Exception) {
+            Log.e(TAG, "[ConsumptionReporting] onException of createRetrofitForConsumpReport")
+        }
+    }
+
+    fun startConsumptionReportTimer(clientId: Int) {
+        val timer = Timer()
+        clientsSessionData[clientId]?.serviceAccessInformationRequestTimer = timer
```

Comment:
why do you overwrite the serviceAccessInformationRequestTimer? This timer is used to re-request the Service Access Information

Reply:
It's a typo and it should be consumptionReportingTimer, thanks for finding the bug.  Corrected

Reply:
I will address the remaining issues in the changes I am currently making. As this PR is already merged please do not add any more changes at this point as they will be lost (see also my comment at the bottom)